### PR TITLE
Use S3 to serve static

### PR DESCRIPTION
**Pull request by mbeacom**
##### Description

The goal of this PR is to migrate from whitenoise to S3 for serving static.  Requests will hit our CDN and pull from S3 opposed to whitenoise.

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)
static

##### Testing

Locally

### CODE REVIEW

**Comment by owocki** [comment]
didnt test but code seems fine to me.  

does this mean we have to upload files to S3 upon every deploy?  or do they hit s3 then our cdn and then cache to s3 upon cache miss?  maybe i need to read the docs :P

agree, lets test on stage

**Comment by owocki** [comment]
one wrinkle is how we're going to do cache invalidations for things like avatars

**Comment by mbeacom** [comment]
@owocki Cache invalidation for avatars shouldn't be an issue.  We're salting the avatar asset any time it's updated/changed to a new asset, so it *should* always serve the latest asset.  This will additionally serve any user assets (media like avatars) from it's own subdomain and static from a separate subdomain.
As for static asset updates... `collectstatic` will automatically upload assets that are new or have changed since the last time it was ran.  `collectfast` uses the DB cache backend to store asset names and hashes, so it should only upload assets when they change.  `collectstatic` is a *bit* slower than it is when using `whitenoise`, but it's nothing near the time taken when we were using `brotlipy`.

We're still using `Manifest` based static files, so static assets are still hashed and then uploaded.  We *could* introduce a hard hash key by deployment, but that would mean we're updating every asset anytime we adjust the fixed hash, which would lead to more time spent performing collectstatic and even more bandwidth usage for return users.

**Comment by owocki** [comment]
> As for static asset updates... collectstatic will automatically upload assets that are new or have changed since the last time it was ran.

sounds good.. as long as this doesnt massively increase deploy times (esp since our site 502s during deploys)

File: app/app/urls.py [hunk 1]
```diff
@@ -160,10 +160,10 @@
     # images
     re_path(r'^funding/embed/?', dashboard.embed.embed, name='embed'),
     re_path(r'^funding/avatar/?', avatar.views.handle_avatar, name='avatar'),
-    re_path(r'^static/avatar/(.*)/(.*)?', avatar.views.handle_avatar, name='org_avatar'),
```
Comment (owocki) [inline app/app/urls.py]: why were these changed?  

the avatar / profile pic lookups are among the most comment requests we get... moving it out of the `/static/` dir is going to massively increase the number of these requests that hit the main app server

File: app/app/urls.py [hunk 2]
```diff
@@ -160,10 +160,10 @@
...
-    re_path(r'^static/viz/graph/(.*)?$', dataviz.d3_views.viz_graph, name='viz_graph'),
-    re_path(r'^static/viz/sscatterplot/(.*)?$', dataviz.d3_views.viz_scatterplot_stripped, name='viz_sscatterplot'),
```
Comment (owocki) [inline app/app/urls.py]: same here... these requests are pretty performance intensive.. thats why they were cached.

they're not really dynamic to the extent that we need to worry about them busting the cache either

**Comment by owocki** [comment]
there are still hardcoded relative `/static` paths all over the codebase....


```
/Users/kevinowocki/Sites/gitcoin/web/app/assets/onepager/css/main.css:
    8  @font-face {
    9    font-family: 'futura-pt';
   10:   src: url('/static/v2/fonts/futura/font4.ttf')/* TTF file for CSS3 browsers */
   11  }
   12  
   ..
   15  @font-face {
   16    font-family: 'Muli';
   17:   src: url('/static/v2/fonts/Muli/Regular.ttf');
   18    font-weight: 400;
   19  }
   ..
   28    color: #000000;
   29    background-color: #0d023b;
   30:   background-image: url('/static/onepager/images/overlay.png'), -moz-linear-gradient(60deg, rgba(255, 165, 150, 0.5) 5%, rgba(0, 228, 255, 0.35)), url('/static/v2/images/header-bg.png');
   31:   background-image: url('/static/onepager/images/overlay.png'), -webkit-linear-gradient(60deg, rgba(255, 165, 150, 0.5) 5%, rgba(0, 228, 255, 0.35)), url('/static/v2/images/header-bg.png');
   32:   background-image: url('/static/onepager/images/overlay.png'), -ms-linear-gradient(60deg, rgba(255, 165, 150, 0.5) 5%, rgba(0, 228, 255, 0.35)), url('/static/v2/images/header-bg.png');
   33:   background-image: url('/static/onepager/images/overlay.png'), linear-gradient(60deg, rgba(255, 165, 150, 0.5) 5%, rgba(0, 228, 255, 0.35)), url('/static/v2/images/header-bg.png');
   34    background-repeat: no-repeat;
   35    background-size: 110%;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/about.css:
  112  .gitcoin-svg {
  113    background-size: cover !important;
  114:   background-image: url("/static/v2/images/ios/gitcoin.svg");
  115    width: 2em;
  116    height: 2em;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/avatar_builder.css:
   36  
   37  #avatar-builder__sections li.complete {
   38:   list-style-image: url('/static/v2/images/check-dark.svg');
   39  }
   40  
   ..
  151  
  152  .avatar-option.selected::after {
  153:   content: url('/static/v2/images/check.svg');
  154    background-color: #150069;
  155  }

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/base.css:
  578  textarea.loading,
  579  input[type=text].loading {
  580:   background-image: url('/static/v2/images/loading_v2.gif');
  581    background-repeat: no-repeat;
  582    background-position: right;
  ...
  698  
  699  .alert .message::before {
  700:   content: url('/static/v2/images/info.svg');
  701    margin-right: 10px;
  702  }
  ...
  712  
  713  .alert.error .message::before {
  714:   content: url('/static/v2/images/close.svg');
  715    height: 30px;
  716  }
  ...
  721  
  722  .alert.success .message::before {
  723:   content: url('/static/v2/images/success.svg');
  724    height: 30px;
  725  }
  ...
  730  
  731  .alert.info .message::before {
  732:   content: url('/static/v2/images/info.svg');
  733    height: 30px;
  734  }
  ...
  739  
  740  .alert.warning .message::before {
  741:   content: url('/static/v2/images/warning.svg');
  742    height: 28px;
  743  }
  ...
  757  
  758  .closebtn.purple {
  759:   content: url('/static/v2/images/close-purple.svg');
  760  }
  761  
  ...
  909  
  910  .dashboard .w-75 .github-comment {
  911:   background: url('/static/v2/images/icon-msg-16.png') no-repeat left;
  912    padding-left: 18px;
  913    margin-right: 10px;
  ...
 1019  
 1020  .ui-selectmenu-icon.ui-icon {
 1021:   background: url('/static/v2/images/caret-down.png') !important;
 1022    background-repeat: no-repeat;
 1023    background-position: 0 !important;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/gitcoin.css:
    3  @font-face {
    4    font-family: 'futura-pt';
    5:   src: url('/static/v2/fonts/futura/font4.ttf')/* TTF file for CSS3 browsers */
    6  }
    7  
    .
   10  @font-face {
   11    font-family: 'OpenSans';
   12:   src: url('/static/v2/fonts/OpenSans/Light.ttf');
   13    font-weight: 300;
   14  }
   ..
   18  @font-face {
   19    font-family: 'OpenSans';
   20:   src: url('/static/v2/fonts/OpenSans/Regular.ttf');
   21    font-weight: 400;
   22  }
   ..
   26  @font-face {
   27    font-family: 'OpenSans';
   28:   src: url('/static/v2/fonts/OpenSans/SemiBold.ttf');
   29    font-weight: 600;
   30  }
   ..
   34  @font-face {
   35    font-family: 'OpenSans';
   36:   src: url('/static/v2/fonts/OpenSans/Bold.ttf');
   37    font-weight: 700;
   38  }
   ..
   42  /* @font-face {
   43    font-family: 'Muli';
   44:   src: url('/static/v2/fonts/Muli/ExtraLight.ttf');
   45    font-weight: 200;
   46  } */
   ..
   50  /* @font-face {
   51    font-family: 'Muli';
   52:   src: url('/static/v2/fonts/Muli/ExtraLightItalic.ttf');
   53    font-weight: 200;
   54    font-style: italic
   ..
   59  @font-face {
   60    font-family: 'Muli';
   61:   src: url('/static/v2/fonts/Muli/Light.ttf');
   62    font-weight: 300;
   63  }
   ..
   67  /* @font-face {
   68    font-family: 'Muli';
   69:   src: url('/static/v2/fonts/Muli/LightItalic.ttf');
   70    font-weight: 300;
   71    font-style: italic;
   ..
   76  @font-face {
   77    font-family: 'Muli';
   78:   src: url('/static/v2/fonts/Muli/Regular.ttf');
   79    font-weight: 400;
   80  }
   ..
   84  /* @font-face {
   85    font-family: 'Muli';
   86:   src: url('/static/v2/fonts/Muli/RegularItalic.ttf');
   87    font-weight: 400;
   88    font-style: italic;
   ..
   93  @font-face {
   94    font-family: 'Muli';
   95:   src: url('/static/v2/fonts/Muli/SemiBold.ttf');
   96    font-weight: 600;
   97  }
   ..
  101  /* @font-face {
  102    font-family: 'Muli';
  103:   src: url('/static/v2/fonts/Muli/SemiBoldItalic.ttf');
  104    font-weight: 600;
  105    font-style: italic;
  ...
  110  /* @font-face {
  111    font-family: 'Muli';
  112:   src: url('/static/v2/fonts/Muli/Bold.ttf');
  113    font-weight: 700;
  114  } */
  ...
  118  /* @font-face {
  119    font-family: 'Muli';
  120:   src: url('/static/v2/fonts/Muli/BoldItalic.ttf');
  121    font-weight: 700;
  122    font-style: italic;
  ...
  172  
  173  .header {
  174:   background-image: url('/static/v2/images/header-bg.png');
  175    background-color: #0d023b;
  176    background-size: cover;
  ...
  182  
  183  .header-contributor {
  184:   background-image: url('/static/v2/images/landing/contributor/hero_header_bg.jpg');
  185    background-color: #3f00ff;
  186  }
  187  
  188  #gitcoin-background {
  189:   background-image: url("/static/v2/images/header-bg.png");
  190    background-size: 2000px;
  191    background-position: 0 0;
  ...
  330  .light-bg {
  331    background-color: rgba(239, 243, 238, 0.7);
  332:   background-image: url('/static/v2/images/light-bg.png');
  333    background-size: contain;
  334    background-position: bottom;
  ...
  338  .white-light-bg {
  339    background-color: white;
  340:   background-image: url('/static/v2/images/light-bg.png');
  341    background-size: contain;
  342    background-position: bottom;
  ...
  430  
  431  .footer__main {
  432:   background-image: url('/static/v2/images/footer.jpg');
  433    background-color: #0d023b;
  434    background-size: contain;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/ios.css:
   88  
   89  .ios .video__wrapper .playpause {
   90:   background-image: url('/static/v2/images/ios/play.png');
   91    background-repeat: no-repeat;
   92    width: 50%;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/landing_page.css:
  360    padding-left: 4rem;
  361    margin-bottom: 0;
  362:   list-style-image: url("/static/v2/images/check_green.svg");
  363  }
  364  

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/leaderboard.css:
   73  
   74  #leaderboard .winner-content .winner-position-first {
   75:   background-image: url('/static/v2/images/leaderboard-1.png');
   76  }
   77  
   78  #leaderboard .winner-content .winner-position-second {
   79:   background-image: url('/static/v2/images/leaderboard-2.png');
   80  }
   81  
   82  #leaderboard .winner-content .winner-position-third {
   83:   background-image: url('/static/v2/images/leaderboard-3.png');
   84  }
   85  

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/profile.css:
  156  
  157  .profile-bounties--loading {
  158:   background-image: url('/static/v2/images/loading_v2.gif');
  159    background-repeat: no-repeat;
  160    background-position: 0 -1rem;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/vision.css:
   22    display: grid;
   23    background-color: var(--tl-bg);
   24:   background-image:url(/static/v2/images/vision/illu.png);
   25    background-repeat: no-repeat;
   26    background-position: center bottom;
   ..
   90  
   91  li.consensys {
   92:   list-style-image: url(/static/v2/images/vision/consensys.svg);
   93  }
   94  li.gitcoin {
   95:   list-style-image: url(/static/v2/images/vision/gitcoin.svg);
   96  }
   97  li.macro {
   98:   list-style-image: url(/static/v2/images/vision/macro.svg);
   99  }
  100  li.market {
  101:   list-style-image: url(/static/v2/images/vision/market.svg);
  102  }
  103  

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/external_bounties/bounties.css:
   11  
   12  .body {
   13:   background-image: url('/static/v2/images/background.png');
   14    background-repeat: repeat;
   15  }

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/external_bounties/bounty.css:
    5  
    6  .body {
    7:   background-image: url('/static/v2/images/background.png');
    8    background-repeat: repeat;
    9  }

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/forms/checkbox.css:
   17    border: 1px solid #dbdbdb;
   18    border-radius: 2px;
   19:   content: url('/static/v2/images/check.svg');
   20    display: inline-flex;
   21    justify-content: center;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/forms/input.css:
   23    padding-right: 2.8rem;
   24    color: #D50000;
   25:   background-image: url('/static/v2/images/close-red.png');
   26    background-repeat: no-repeat;
   27    background-size: auto 1.313rem;
   ..
   31  .form__input.valid {
   32    padding-right: 2.8rem;
   33:   background-image: url('/static/v2/images/success.png');
   34    background-repeat: no-repeat;
   35    background-size: auto 1.313rem;

/Users/kevinowocki/Sites/gitcoin/web/app/assets/v2/css/forms/select.css:
   96  
   97  .form__select2 .select2-container--default .select2-selection--single .select2-selection__arrow::after {
   98:   background-image: url('/static/v2/images/arrow-down.png');
   99    background-size: contain;
  100    content: '';

/Users/kevinowocki/Sites/gitcoin/web/app/dashboard/notifications.py:
  296              try:
  297                  headers = {'Content-Type': 'application/json'}
  298:                 body = {"content": msg, "avatar_url": "https://gitcoin.co/static/v2/images/helmet.png"}
  299                  discord_response = requests.post(
  300                      subscriber.discord_webhook_url, headers=headers, json=body

/Users/kevinowocki/Sites/gitcoin/web/app/dashboard/templates/ens/base.html:
    4    <head>
    5      <title>{% trans "ENS Subdomains | Gitcoin" %}</title>
    6:     {% include 'shared/cards_pic.html' with title='Gitcoin | _you_.gitcoin.eth ENS Subdomains'  card_desc='*NEW* Register `_you_.gitcoin.eth` with one click for free' avatar_url="https://gitcoin.co/static/v2/images/ens_gitcoin.png"  %}
    7      <meta charset="utf-8" />
    8      <meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=no">

/Users/kevinowocki/Sites/gitcoin/web/app/dataviz/templates/dataviz/circles.html:
   79  
   80  var url = document.location.href + '?data=1&format=json';
   81: //var url = '/static/v2/js/stats/data.json'
   82  d3.json(url, function(error, root) {
   83    if (error) throw error;

/Users/kevinowocki/Sites/gitcoin/web/app/dataviz/templates/dataviz/draggable.html:
    5  <meta charset="utf-8">
    6    <title>User balance progression over time</title>
    7:   <link href="/static/v2/js/dataviz/dragit.css" rel="stylesheet">
    8:   <script src="/static/v2/js/dataviz/d3.v3.js"></script>
    9:   <script src="/static/v2/js/dataviz/dragit.js"></script>
   10  </head>
   11  <title>User balance progression over time</title>

/Users/kevinowocki/Sites/gitcoin/web/app/dataviz/templates/dataviz/heatmap.html:
   87      <script type="text/javascript" src="//d3js.org/d3.v3.js"></script>
   88      <script src="{% static "v2/js/jquery.js" %}"></script>
   89:     <link rel="stylesheet" href="/static/v2/js/dataviz/heatmap.css">
   90    </head>
   91    <body>
   ..
  118      </div>
  119  
  120:     <script type="text/javascript" src="/static/v2/js/dataviz/heatmap.js"></script>
  121    </body>
  122  </html>

/Users/kevinowocki/Sites/gitcoin/web/app/dataviz/templates/dataviz/scatterplot_stripped.html:
   23  
   24  </style>
   25: <script src="/static/v2/js/jquery.js"></script>
   26  <script src="//d3js.org/d3.v3.min.js"></script>
   27  <script>

/Users/kevinowocki/Sites/gitcoin/web/app/dataviz/templates/dataviz/spiral.html:
  104      {% endfor %}
  105      </script>
  106:     <script type="text/javascript" src="/static/v2/js/dataviz/spiral.js"></script>
  107    </body>
  108  

/Users/kevinowocki/Sites/gitcoin/web/app/dataviz/templates/dataviz/sunburst.html:
   39      <link rel="stylesheet" type="text/css"
   40        href="https://fonts.googleapis.com/css?family=Open+Sans:400,600">
   41:     <link rel="stylesheet" type="text/css" href="/static/v2/js/dataviz/sequences.css"/>
   42    </head>
   43  
   ..
  100  </script>
  101  
  102: <script type="text/javascript" src="/static/v2/js/dataviz/sequences.js"></script>
  103  <script type="text/javascript">
  104    // Hack to make this example display correctly in an iframe on bl.ocks.org

/Users/kevinowocki/Sites/gitcoin/web/app/external_bounties/models.py:
   86              i = self.pk % len(icons)
   87              icon = icons[i]
   88:             return f'{settings.BASE_URL}static/v2/images/icons/{icon}'
   89  
   90      @property

/Users/kevinowocki/Sites/gitcoin/web/app/marketing/utils.py:
   80  
   81      if not icon_url:
   82:         icon_url = 'https://gitcoin.co/static/v2/images/helmet.png'
   83  
   84      try:
   ..
  130  
  131      if not icon_url:
  132:         icon_url = 'https://gitcoin.co/static/v2/images/helmet.png'
  133  
  134      try:

/Users/kevinowocki/Sites/gitcoin/web/js-sdk/src/index.js:
   33      anchor.appendChild(logo);
   34      // need a srcset or svg logo for responsiveness
   35:     logo.src = 'https://s.gitcoin.co/static/v2/images/logo.png';
   36      logo.className = 'gitcoin-widget__logo';
   37      return anchor;

73 matches across 27 files

File: app/app/urls.py (hunk 1 above)
Comment (mbeacom) [reply app/app/urls.py]: @owocki I updated the cloudfront caching entries from `/static` to `/dynamic` since we're no longer serving static from the same cloudfront distribution.  We shouldn't see any change on this endpoints beyond the initial surge to re-cache the assets as necessary.

File: app/app/urls.py (hunk 2 above)
Comment (mbeacom) [reply app/app/urls.py]: @owocki I renamed them to `dynamic` because they aren't truly static assets or endpoints.  They're still generated dynamically on some interval.  The CDN entries were updated before this went live to accommodate the change.  We shouldn't experience any performance hit beyond the initial recaching on the new path.

**Comment by mbeacom** [comment]
I suppose I shouldn't have assumed we were all using the `static` helpers in new code... apologies... will update where relevant.  Some of the entries in the above snippet are referencing `s.gitcoin.co`, but I see your point.

File: app/app/urls.py (hunk 2 above)
Comment (owocki) [reply app/app/urls.py]: word; i see the dynamic endpoint. makes sense

**Comment by owocki** [comment]
>  As for the .css static hardcoding, those should be handled during collectstatic and automatically replaced with the proper paths

ahh got it; did not know that
